include/uk/plat/memory.h

```diff
@@ -132,7 +132,15 @@ ukplat_memregion_find_next(int i, __u32 type, __u32 flags, __u32 fmask,
 {
 	struct ukplat_memregion_desc *desc;
 	__u32 stype, sflags;
-	int rc;
+	int rc, count;
+    	count = ukplat_memregion_count();
+
+	/* Sanity check before to get a memory region using ukplat_memregion_get
+	 * this prevent  Initialize memory allocator in platform linuxu 
+	 * */
+    	if(i >= count){
```

Comment:
I believe this check is already done by `ukplat_memregion_get`, isn't it?

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/kvm/x86/memory.c#L26

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/kvm/arm/memory.c#L173

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/xen/memory.c#L143

https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/linuxu/memory.c#L185

Reply:
perhaps the problem is that  the function ukplat_memregion_count not only count the number of memregions. this function also initialize some heap and I dont see that call here: [unikraft/plat/linuxu/memory.c](https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/linuxu/memory.c#L185)
~~~~
if (!have_heap) {
		rc = __linuxu_plat_heap_init();
		have_heap = (rc == 0) ? 1 : 0;
	}

	if (!have_initrd) {
		rc = __linuxu_plat_initrd_init();
		have_initrd = (rc == 0) ? 1 : 0;
	}
~~~~

Reply:
I somehow just now received your reply, yes, indeed that is why the fix works. Not because of the sanity check.

Reply:
sorry, I also get confused. and yes, you are correct.
would it work if  add the call  ukplat_memregion_count as in [unikraft/plat/xen/memory.c](https://github.com/unikraft/unikraft/blob/4c7352c0a5ad3a2785fb6aec3043a7c1f46e7883/plat/xen/memory.c#L143)?

I can work in the changes, So, please let me know what would be the best way to do it. 

Thank you.

Reply:
Yeah, that would look good I guess. But it would be a good idea, in my opinion, to also add a check at the beginning of `ukplat_memregion_count()` for the `static` variables, to avoid re-initialization attempts if one of them is 0. This is how I would see the function, since if you successfully went through it once, I guess you do not need to repeat the run (I might be wrong here, @marcrittinghaus might have more context).
```C
int ukplat_memregion_count(void)
{
	static int init = 0;
	int have_initrd = 0;
	int have_heap = 0;
	int rc = 0;

	if (init)
		return init;
	/*
	 * NOTE: The heap size and initrd file can be changed by a
	 * library parameter. We assume that those ones are processed
	 * by the boot library shortly before memory regions are
	 * scanned. This is why we initialize the heap here.
	 */
	rc = __linuxu_plat_heap_init();
	have_heap = (rc == 0) ? 1 : 0;

	rc = __linuxu_plat_initrd_init();
	have_initrd = (rc == 0) ? 1 : 0;

	init = have_heap + have_initrd;

	return init;
}
```

Reply:
I say let's wait for the others' opinion as well before making any change :).

Reply:
Sure 😊

Reply:
I agree with @mogasergiu about adding the check